src/KilnUniV3.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+import {KilnBase, GemLike} from "./KilnBase.sol";
+import {TwapProduct}       from "./uniV3/TwapProduct.sol";
+
+// https://github.com/Uniswap/v3-periphery/blob/b06959dd01f5999aa93e1dc530fe573c7bb295f6/contracts/SwapRouter.sol#L132
+interface SwapRouterLike {
+    function exactInput(ExactInputParams calldata params) external returns (uint256 amountOut);
+    function factory() external returns (address factory);
+}
+
+// https://github.com/Uniswap/v3-periphery/blob/b06959dd01f5999aa93e1dc530fe573c7bb295f6/contracts/interfaces/ISwapRouter.sol#L26
+// https://docs.uniswap.org/protocol/guides/swaps/multihop-swaps#input-parameters
+struct ExactInputParams {
+    bytes   path;
+    address recipient;
+    uint256 deadline;
+    uint256 amountIn;
+    uint256 amountOutMinimum;
+}
+
+contract KilnUniV3 is KilnBase, TwapProduct {
+    uint256 public scope; // [Seconds]  Time period for TWAP calculations
+    uint256 public yen;   // [WAD]      Relative part of the TWAP's price to insist on
+    bytes   public path;  //            ABI-encoded UniV3 compatible path
+
+    address public immutable uniV3Router;
+    address public immutable receiver;
+
+    event File(bytes32 indexed what, bytes data);
+
+    // @notice initialize a Uniswap V3 routing path contract
+    // @dev TWAP-relative trading is enabled by default. Fire will perform the trade
+    //      only when the amount of tokens received is better than the 1 hour average price.
+    //      file("yen", 0) to disable TWAP trading and accept any output amount.
+    // @param _sell the contract address of the token that will be sold
+    // @param _buy  the contract address of the token that will be purchased
+    // @param _uniV3Router the address of the current Uniswap V3 swap router
+    // @param _receiver the address of the account which will receive the funds to be bought
+    constructor(
+        address _sell,
+        address _buy,
+        address _uniV3Router,
+        address _receiver
+    )
+        KilnBase(_sell, _buy)
+        TwapProduct(SwapRouterLike(_uniV3Router).factory())
+    {
+        uniV3Router = _uniV3Router;
+        receiver    = _receiver;
+
+        scope = 1 hours;
```

Comment:
Why not make this manageable or at least as constructor param?

Reply:
It is fileable.
The default values for `scope` and `yen` are not mandatory, but are a safety measure (as leaving them 0 by mistake would work, but with a more risky configuration).

Reply:
Yes sorry, just got confused with the comment in the constructor and didn't check it well.
"Fire will perform the trade only when the amount of tokens received is better than the 1 hour average price."
Maybe I'd changing it to something like: "With the initial values, fire will perform the trade only when the amount of tokens received is equal or better than the 1 hour average price."

Reply:
changed on [e8c0964](https://github.com/makerdao/dss-kiln/pull/9/commits/e8c0964fe1a80e828163e3aad9a8caaf22e3fc9e)